scene/main/canvas_item.cpp

```diff
@@ -509,40 +509,49 @@ void CanvasItem::draw_rect(const Rect2 &p_rect, const Color &p_color, bool p_fil
 		}
 
 		RenderingServer::get_singleton()->canvas_item_add_rect(canvas_item, p_rect, p_color);
+	} else if (p_width >= p_rect.size.width || p_width >= p_rect.size.height) {
+		Size2 pos = p_rect.position - Size2(0.5 * p_width, 0.5 * p_width);
+		Size2 size = p_rect.size + Size2(p_width, p_width);
+		RenderingServer::get_singleton()->canvas_item_add_rect(canvas_item, Rect2(pos, size), p_color);
 	} else {
-		// Thick lines are offset depending on their width to avoid partial overlapping.
-		// Thin lines don't require an offset, so don't apply one in this case
-		real_t offset;
-		if (p_width >= 2) {
-			offset = p_width / 2.0;
-		} else {
-			offset = 0.0;
+		Size2 pos = p_rect.position;
+		Size2 size = p_rect.size;
+		int width = (int)Math::round(p_width);
+		real_t offset = 0.5f * width;
+
+		if (width % 2) {
+			pos += Size2(0.5, 0.5);
+			size -= Size2(1, 1);
 		}
```

Comment:
Actually, I would like someone to review this part. I'm not sure if this is the correct solution. Especially considering `CanvasItem` and/or `Viewport` scaling.

In fact, the `p_width` type needs to be replaced from `float` to `int`, since it is rounded. I just don't know how else to make the line appear exactly the width that is specified. Therefore, I take into account the width parity, in order to snap the lines to the pixel grid.

Reply:
I removed this part.